File: web-next/lib/url-policy.ts
const PROD_ENVS = new Set(["production", "prod", "staging", "stage"]);
const VALID_POLICIES = new Set(["auto", "force_http", "force_https"]);

const normalize = (value: string | undefined, fallback: string): string => {
  const trimmed = (value || "").trim().toLowerCase();
  return trimmed || fallback;
};

const isPrivateIpv4 = (host: string): boolean => {
  const rawParts = host.split(".");
  if (rawParts.length !== 4) {
    return false;
  }
  if (!rawParts.every((part) => /^\d{1,3}$/.test(part))) {
    return false;
  }
  const parts = rawParts.map((part) => Number.parseInt(part, 10));
  if (parts.some((part) => part < 0 || part > 255)) {
    return false;
  }
  if (parts[0] === 10) return true;
  if (parts[0] === 127) return true;
  if (parts[0] === 192 && parts[1] === 168) return true;
  if (parts[0] === 172 && parts[1] >= 16 && parts[1] <= 31) return true;
  return false;
};

export const isLocalOrPrivateHost = (host: string): boolean => {
  const hostname = host.trim().toLowerCase();
  if (!hostname) return false;
  if (hostname === "localhost" || hostname === "0.0.0.0" || hostname === "[::1]") {
    return true;
  }
  if (hostname.endsWith(".localhost") || hostname.endsWith(".local")) {
    return true;
  }
  return isPrivateIpv4(hostname);
};

export const getUrlSchemePolicy = (): "auto" | "force_http" | "force_https" => {
  const value = normalize(
    process.env.NEXT_PUBLIC_URL_SCHEME_POLICY || process.env.URL_SCHEME_POLICY,
    "auto"
  );
  if (VALID_POLICIES.has(value)) {
    return value as "auto" | "force_http" | "force_https";
  }
  return "auto";
};

export const getRuntimeEnv = (): string => {
  return normalize(process.env.NEXT_PUBLIC_ENV || process.env.NODE_ENV, "development");
};

export const resolveHttpScheme = (hostname: string): "http" | "https" => {
  const policy = getUrlSchemePolicy();
  if (policy === "force_http") return "http";
  if (policy === "force_https") return "https";
  if (isLocalOrPrivateHost(hostname)) return "http";
  return PROD_ENVS.has(getRuntimeEnv()) ? "https" : "http";
};

export const buildHttpBaseUrl = (hostname: string, port?: number): string => {
  const scheme = resolveHttpScheme(hostname);
  return port ? `${scheme}://${hostname}:${port}` : `${scheme}://${hostname}`;
};

export const applyHttpPolicyToUrl = (value: string): string => {
  try {
    const parsed = new URL(value);
    if (parsed.protocol !== "http:" && parsed.protocol !== "https:") {
      return value;
    }
    parsed.protocol = `${resolveHttpScheme(parsed.hostname)}:`;
    return parsed.toString().replace(/\/$/, "");
  } catch {
    return value;
  }
};

File: venom_core/utils/url_policy.py
"""Central policy for resolving HTTP scheme across environments."""

from __future__ import annotations

import ipaddress
from urllib.parse import urlparse, urlunparse

from venom_core.config import SETTINGS

VALID_URL_SCHEME_POLICIES = {"auto", "force_http", "force_https"}
PROD_ENVS = {"production", "prod", "staging", "stage"}


def normalize_url_scheme_policy(policy: str | None = None) -> str:
    value = (policy or getattr(SETTINGS, "URL_SCHEME_POLICY", "auto") or "auto").strip()
    lowered = value.lower()
    if lowered in VALID_URL_SCHEME_POLICIES:
        return lowered
    return "auto"


def normalize_env(env: str | None = None) -> str:
    value = (env or getattr(SETTINGS, "ENV", "development") or "development").strip()
    return value.lower()


def is_local_or_private_host(hostname: str) -> bool:
    host = (hostname or "").strip().lower()
    if not host:
        return False
    if host in {"localhost", "0.0.0.0", "::1"}:
        return True
    if host.endswith(".localhost"):
        return True
    if host.endswith(".local"):
        return True
    try:
        ip = ipaddress.ip_address(host)
    except ValueError:
        return False
    return ip.is_loopback or ip.is_private or ip.is_link_local


def resolve_http_scheme(
    hostname: str, env: str | None = None, policy: str | None = None
) -> str:
    active_policy = normalize_url_scheme_policy(policy)
    if active_policy == "force_https":
        return "https"
    if active_policy == "force_http":
        return "http"

    if is_local_or_private_host(hostname):
        return "http"

    return "https" if normalize_env(env) in PROD_ENVS else "http"


def apply_http_policy_to_url(
    url: str, env: str | None = None, policy: str | None = None
) -> str:
    parsed = urlparse(url)
    if parsed.scheme not in {"http", "https"} or not parsed.hostname:
        return url

    scheme = resolve_http_scheme(parsed.hostname, env=env, policy=policy)
    return urlunparse(parsed._replace(scheme=scheme))


def build_http_url(
    hostname: str,
    port: int | None = None,
    path: str = "",
    env: str | None = None,
    policy: str | None = None,
) -> str:
    scheme = resolve_http_scheme(hostname, env=env, policy=policy)
    netloc = f"{hostname}:{port}" if port else hostname
    normalized_path = path if not path or path.startswith("/") else f"/{path}"
    return urlunparse((scheme, netloc, normalized_path, "", "", ""))
